mri/operators/gradient/gradient.py

```diff
@@ -0,0 +1,165 @@
+        input data array.
+    fourier_op: instance
+        a Fourier operator instance.
+    verbose: int, default 0
+        Debug verbosity. Prints debug information during initialization if 1.
+    """
+    def __init__(self, data, fourier_op, verbose=0, **kwargs):
+        if fourier_op.n_coils != 1:
+            super(GradAnalysis, self).__init__(data, fourier_op.op,
+                                               fourier_op.adj_op,
+                                               (fourier_op.n_coils,
+                                                *fourier_op.shape),
+                                               verbose=verbose,
+                                               **kwargs)
+        else:
+            super(GradAnalysis, self).__init__(data, fourier_op.op,
+                                               fourier_op.adj_op,
+                                               fourier_op.shape,
+                                               verbose=verbose,
+                                               **kwargs)
+        self.fourier_op = fourier_op
+
+
+class GradSynthesis(GradBaseMRI):
+    """ Gradient Synthesis class.
+    This class defines the grad operators for:
+            (1/2) * sum(||Ft Psi_t alpha - yl||^2_2,l)
+    Attributes
+    ----------
+    data: np.ndarray
+        input 2D data array.
+    fourier_op: instance
+        a Fourier operator instance.
+    linear_op: instance
+        a linear operator instance.
+    verbose: int, default 0
+        Debug verbosity. Prints debug information during initialization if 1.
+    """
+    def __init__(self, data, linear_op, fourier_op, verbose=0, **kwargs):
+        self.fourier_op = fourier_op
+        self.linear_op = linear_op
+        coef = linear_op.op(np.squeeze(np.zeros((linear_op.n_coils,
+                                                 *fourier_op.shape))))
+        self.linear_op_coeffs_shape = coef.shape
+        super(GradSynthesis, self).__init__(data,
+                                            self._op_method,
+                                            self._trans_op_method,
+                                            self.linear_op_coeffs_shape,
+                                            verbose=verbose,
+                                            **kwargs)
+
+    def _op_method(self, data):
+        return self.fourier_op.op(self.linear_op.adj_op(data))
+
+    def _trans_op_method(self, data):
+        return self.linear_op.op(self.fourier_op.adj_op(data))
+
+
+class GradSelfCalibrationAnalysis(GradBaseMRI):
+    """ Gradient Analysis class for parallel MR reconstruction based on the
+    coil sensitivity profile.
+    This class defines the grad operators for:
+            (1/2) * sum(||Ft Sl x - yl||^2_2,l)
+    Attributes
+    ----------
+    data: np.ndarray
+        input observed data array.
+    fourier_op: instance
+        a Fourier operator instance.
+    Smaps: np.ndarray
+        Coil sensitivity profile [L, Nx, Ny, Nz]
```

Comment:
Does it have to be 3D?

Reply:
Nope, L x image shape? How do we write it?

Reply:
`[L, *data_shape]` ?

Reply:
Done